src/config.rs

```diff
@@ -1,33 +1,140 @@
-use std::path::{Path, PathBuf};
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::HashMap,
+    fs::File,
+    path::{Path, PathBuf},
+    process,
+};
+
+/// Restart conditions for a service.
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all = "lowercase")]
+enum Restart {
+    Never,
+    Always,
+    Unexpected,
+}
 
-#[derive(Debug)]
 /// Configuration structure used to run a task.
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(deny_unknown_fields)]
+#[serde(default)]
 pub struct Config {
     /// Command to run.
     cmd: String,
     /// Number of processors this task should run with.
     // u16 is fine because we are not expecting a machine to run more
     // than 2^16 proc at any single time.
     numprocs: u16,
+    /// Program mask running under
+    umask: i32,
     /// Working directory of the task.
     workingdir: PathBuf,
+    /// Service starting at load time.
+    autostart: bool,
+    /// Service restarting when finished.
+    autorestart: Restart,
+    /// Autorestart callback trigger value.
+    exitcodes: Vec<i32>,
+    /// Restart attempts.
+    startretries: i32,
+    /// Delay before start.
+    starttime: i32,
+    /// Signal received by supervisord trigger service termination.
+    stopsignal: String,
+    /// Running period.
+    stoptime: i32,
+    /// Stdout logfile path.
+    stdout: PathBuf,
+    /// Stderr logfile path.
+    stderr: PathBuf,
+    /// Environment variable service is running into.
+    env: HashMap<String, String>,
 }
 
 impl Default for Config {
     fn default() -> Self {
+        let bin = "ls";
+        let home = dirs::home_dir();
+        if home.is_none() {
+            eprintln!("Impossible to get user home directory!");
+            process::exit(1);
+        }
+        let home = home.unwrap();
+
+        #[cfg(debug_assertions)]
+        println!("{:?}", home);
+
         Self {
-            cmd: "ls".into(),
+            cmd: bin.into(),
             numprocs: 1,
-            workingdir: PathBuf::from("/tmp"),
+            umask: 0o027,
+            workingdir: home.clone(),
+            autostart: true,
+            autorestart: Restart::Never,
+            exitcodes: {
+                let mut v = Vec::new();
+                v.push(0);
+                v
+            },
+            startretries: 0,
+            starttime: 0,
+            stopsignal: "SIGTERM".to_string(),
+            stoptime: 0,
+            stdout: home.join(bin.to_string() + ".stdout"), // Would be lovely to replace "bin" by "cmd" from the struct itself
+            stderr: home.join(bin.to_string() + ".stderr"), // Same
+            env: HashMap::new(),
         }
     }
 }
 
-impl std::convert::TryFrom<&Path> for Config {
-    type Error = std::io::Error;
+/// # Errors
+///
+/// Will return `Err` if `file` does not exist or the user does not have
+/// permission to read it or yaml format is not respected.
+pub fn parse(
```

Comment:
Wondering why you removed the `TryFrom` and instead used a user-defined `parse()` function.
It is usually more idiomatic to either call `Config::new` or `Config::try_from` rather than calling `parse()` which doesn't specify what struct / object it will create.
I think it would be better to use either `TryFrom` or `new` :)

Reply:
Because of incompatible types. I do not know how to fix this:
```
error[E0053]: method `try_from` has an incompatible type for trait
   --> src/config.rs:94:5
    |
94  | /     fn try_from(
95  | |         file_name: &Path,
96  | |     ) -> Result<HashMap<String, HashMap<String, Self>>, serde_yaml::Error> {
97  | |         let file = File::open(&file_name).expect("Unable to open config file");
98  | |         let d: HashMap<String, HashMap<String, Config>> = serde_yaml::from_reader(file)?;
99  | |         Ok(d)
100 | |     }
    | |_____^ expected struct `config::Config`, found struct `std::collections::HashMap`
```

Reply:
See [this comment](https://github.com/pscott/TaskMaster/pull/18#discussion_r493738767) :)